main/src/ca/uwaterloo/flix/language/phase/unification/ContextReduction.scala

```diff
@@ -0,0 +1,125 @@
+/*
```

Comment:
I have not reviewed this class, but I will once I have re-read the paper and its implementation has become a bit more "stable".

Reply:
Please ensure that this does not break the compiler :)

Reply:
> Please ensure that this does not break the compiler :)

Do you mean compiler performance?

Reply:
Yes, and throwing random exceptions :)  I think our 9k tests should guard against the latter. (Hopefully).

Reply:
Performance testing is still a little erratic on my end, but overall performance seems reasonable.

Reply:
Can you post some numbers? i am fine with +10% but hopefully not much more than that.

Reply:
Here's the output of a run of each. It shows a huge increase in the Namer, but I don't think that all can be attributed to my changes. I'll do another run to see if there's any consistency.

master throughput | feature throughput | absolute difference | percent difference
--- | --- | --- | ---
20559 | 17671 | -2888 | -14.0%

phase | master time | feature time | absolute difference | percent difference
--- | --- | --- | --- | ---
Reader | 928610 | 970697 | +42087 | +4.5%
Parser | 145208235 | 185335666 | +40127431 | +27.6%
Weeder | 13356042 | 15047511 | +1691469 | +12.7%
Namer | 44660818 | 74541435 | +29880617 | +66.9%
Resolver | 99049949 | 69805188 | -29244761 | -29.5%
Typer | 359022653 | 293551584 | -65471069 | -18.2%
Stratifier | 13902226 | 13250813 | -651413 | -4.7%
PatternExhaustiveness | 27507803 | 29000817 | +1493014 | +5.4%
Redundancy | 24487959 | 26403009 | +1915050 | +7.8%
Safety | 4768498 | 4537286 | -231212 | -4.8%
Documentor | 5186 | 1710 | -3476 | -67.0%
Monomorph | 256046228 | 265731710 | +9685482 | +3.8%
Synthesize | 16385762 | 15434935 | -950827 | -5.8%
Simplifier | 87102063 | 86718147 | -383916 | -0.4%
ClosureConv | 48268090 | 32791059 | -15477031 | -32.1%
LambdaLift | 12246253 | 11915007 | -331246 | -2.7%
Tailrec | 4962953 | 4979260 | +16307 | +0.3%
Optimizer | 11386569 | 11795016 | +408447 | +3.6%
TreeShaker | 50127636 | 47792923 | -2334713 | -4.7%
VarNumbering | 7563882 | 8153825 | +589943 | +7.8%
Finalize | 38581136 | 38083433 | -497703 | -1.3%
JvmBackend | 819125307 | 710658945 | -108466362 | -13.2%

Reply:
OK, take a look at the namer.

The problem with `--Xthroughput-phases` is that it does not compute the median of the runs. I know it is a hassle, but maybe you could fix that? `BenchmarkCompiler:25` and then maintain a map from phaseNames to the median of their execution time.

Reply:
Note that the resolver is much faster, so I really think that these numbers are too unstable.

However, the throughput does suggest that some performance is lost.

Reply:
Second run tells a different story:

master throughput | feature throughput | absolute difference | percent difference
--- | --- | --- | ---
20245 | 21385 | +1140 | +5.6%

phase | master time | feature time | absolute difference | percent difference
--- | --- | --- | --- | ---
Reader | 940944 | 903815 | -37129 | -3.9%
Parser | 147663701 | 143333822 | -4329879 | -2.9%
Weeder | 13855492 | 14168281 | +312789 | +2.3%
Namer | 64076583 | 45081054 | -18995529 | -29.6%
Resolver | 73233529 | 78423303 | +5189774 | +7.1%
Typer | 327463187 | 306137794 | -21325393 | -6.5%
Stratifier | 15223095 | 14628782 | -594313 | -3.9%
PatternExhaustiveness | 29150943 | 28824859 | -326084 | -1.1%
Redundancy | 28914011 | 26220012 | -2693999 | -9.3%
Safety | 5812601 | 4647910 | -1164691 | -20.0%
Documentor | 1724 | 2033 | +309 | +17.9%
Monomorph | 286588209 | 255057491 | -31530718 | -11.0%
Synthesize | 15878285 | 15263025 | -615260 | -3.9%
Simplifier | 88192480 | 88626785 | +434305 | +0.5%
ClosureConv | 33601734 | 33147390 | -454344 | -1.4%
LambdaLift | 12247641 | 12224791 | -22850 | -0.2%
Tailrec | 4871815 | 5017828 | +146013 | +3.0%
Optimizer | 11713196 | 11720348 | +7152 | +0.1%
TreeShaker | 51954554 | 70318725 | +18364171 | +35.3%
VarNumbering | 8217691 | 6068420 | -2149271 | -26.2%
Finalize | 37369380 | 32687282 | -4682098 | -12.5%
JvmBackend | 750393031 | 733507071 | -16885960 | -2.3%

Reply:
Heh... Can you crank up the number of iterations?

Reply:
> compute the median

I will see what I can do

Reply:
From what I can tell, --Xbenchmark-phases only does 1 run. Am I missing something?

Reply:
Yes, that is what you would need to change

Reply:
https://github.com/flix/flix/pull/1508

Reply:
Here's the output with 100 benchmark iterations:

master throughput | feature throughput | absolute difference | percent difference
--- | --- | --- | ---
15521 | 15715 | +194 | +1.2%

phase | master time | feature time | absolute difference | percent difference
--- | --- | --- | --- | ---
Resolver | 70806791 | 70521847 | -284944 | -0.4%
Simplifier | 84723207 | 82764171 | -1959036 | -2.3%
ClosureConv | 34249101 | 33386589 | -862512 | -2.5%
Synthesize | 14484240 | 12821352 | -1662888 | -11.5%
Optimizer | 11951802 | 11534343 | -417459 | -3.5%
Tailrec | 4845288 | 4832185 | -13103 | -0.3%
Finalize | 37438590 | 39520205 | +2081615 | +5.6%
Monomorph | 247336582 | 261499970 | +14163388 | +5.7%
Redundancy | 23925877 | 24901565 | +975688 | +4.1%
Parser | 130506542 | 148993026 | +18486484 | +14.2%
JvmBackend | 701450531 | 707127550 | +5677019 | +0.8%
PatternExhaustiveness | 27097358 | 25107685 | -1989673 | -7.3%
Safety | 4364700 | 4477578 | +112878 | +2.6%
Reader | 297880 | 298606 | +726 | +0.2%
Documentor | 1831 | 1867 | +36 | +2.0%
Typer | 281724635 | 269187355 | -12537280 | -4.5%
LambdaLift | 11546385 | 11289492 | -256893 | -2.2%
Stratifier | 12939165 | 13192415 | +253250 | +2.0%
TreeShaker | 45232759 | 48186194 | +2953435 | +6.5%
Weeder | 12594879 | 13781088 | +1186209 | +9.4%
VarNumbering | 8027236 | 7928162 | -99074 | -1.2%
Namer | 41306351 | 42495903 | +1189552 | +2.9%

Reply:
Great!